### ISSUE TITLE
Authorization issue MSIS7068: Access denied

### ISSUE THREAD

**Issue by Wireliner** [koltyakov/sp-rest-proxy#44]
Good day! 

I started to receive "Access denied" errors for all my REST requests. Couple of days ago everything was working just fine. 

<img width="790" alt="response" src="https://user-images.githubusercontent.com/22619894/36354902-c8000d0c-14fc-11e8-9b9c-df33495dabfb.png">

<img width="793" alt="headers" src="https://user-images.githubusercontent.com/22619894/36354903-c839ae40-14fc-11e8-90fd-080490128441.png">

<img width="565" alt="terminal" src="https://user-images.githubusercontent.com/22619894/36354904-c8726672-14fc-11e8-9a51-2878b89c4c7d.png">

I used version 2.5.0, then updated to 2.5.5 and still have issue. 

Normally, I use my corporate laptop (authenticates automatically) to navigate on SP. But, prefer to develop on my Mac. Recently, I had to request Entrust Security token to be able to login from 3rd party computers. But, I had no issues after that. 

I paste full url (without localhost) into browser and it successfully returns me data in xml format. 

Do you have an idea what could be wrong with authentication? 

Sharepoint Online & Angular 5 project

Regards,
Zhandos

**Comment by koltyakov** [comment]
Hey Jandos,

I guess that this tenant was reconfigured and now using ADFS authentication. Please check auth with ADFS options first. Using AddIn only auth strategy also might be an option.

This thread could be helpful https://github.com/s-KaiNet/node-sp-auth/issues/11.

**Comment by koltyakov** [comment]
Hi @Wireliner, any updates on this? Did you manage to figure out in ADFS options?

**Comment by Wireliner** [comment]
Hi @koltyakov, it is weird but issue has gone, when I connected through corporate network, after switching to private network, the authentication was still ok.

**Comment by koltyakov** [comment]
Yep. Strange. Maybe network proxy settings? Maybe there was an experimental ADFS setup and rollback to SAML after...
Closing the issue. Feel free to reopen if the issue will arise.

**Comment by Wireliner** [comment]
@koltyakov hello! 

Looks like root cause was that company implemented MFA when SP is being access from outside of corporate network. That's why I had no issues when connection at work. Below, is the letter that I recently received. 

<img width="626" alt="mfa" src="https://user-images.githubusercontent.com/22619894/37332522-21f1f344-26d1-11e8-8453-8736e6aef91b.png">

It requires MS Authenticator on mobile device when I login via Browser: 

![img_0239](https://user-images.githubusercontent.com/22619894/37333561-d9eb7f22-26d3-11e8-8aa3-ceffb60bfa51.jpg)

How do you think, would you be able to make MFA authorization work with your rest-proxy?

Regards,
Zhandos

**Comment by koltyakov** [comment]
Hey @Wireliner,

Didn't know you are from Kazakhstan. =) 

MFA is an issue. Not sure that MFA can be treated. 

Can you try OnDemand of AddIn-only auth strategies?

**Comment by Wireliner** [comment]
Salem @koltyakov :)

I tried, but none of options work. Hope you will find solution)

**Comment by koltyakov** [comment]
Does it fail or expect MFA validation? (with OnDemand, with AddIn Only) 
Any error messages?

**Comment by koltyakov** [comment]
Also, out of the interest, can you check device login with [Office 365 CLI](https://github.com/SharePoint/office365-cli), I was thinking of adding this auth mechanism or suggesting adding it to node-sp-auth.

**Comment by koltyakov** [comment]
And, @Wireliner, please double check AddIn Only. I was assured, AddIn Only auth will work even if MFA is activated.

**Comment by Wireliner** [comment]
@koltyakov thanks for your hint! 

Device login with Office 365 CLI didn't work. I had to be admin of tenant, who I'm not, unfortunately. 

For creds auth, I cannot create app secret as I don't have that option in my account. 

But, AddIn only worked! 

What surprised me is that after authorization, when I send REST call to get current user, I get back different credentials (not mine, but of app's that I created). For development purposes it's fine. I can mock current user info. Thank you again! 

<img width="549" alt="currentuser" src="https://user-images.githubusercontent.com/22619894/37408893-426c1108-27c7-11e8-9de3-9b80de2617e6.png">
----
<img width="355" alt="created" src="https://user-images.githubusercontent.com/22619894/37408901-460e891c-27c7-11e8-9543-10e477ae8b2e.png">
----
<img width="261" alt="optimus" src="https://user-images.githubusercontent.com/22619894/37408912-4cb6eb24-27c7-11e8-8bae-3f00a0a69ead.png">
